nixos/modules/security/wrappers/wrapper.c

```diff
@@ -189,15 +201,17 @@ int main(int argc, char **argv) {
     ASSERT('/' == wrapper_dir[0]);
     ASSERT('/' == self_path[len]);
 
-    // Make *really* *really* sure that we were executed as
-    // `self_path', and not, say, as some other setuid program. That
-    // is, our effective uid/gid should match the uid/gid of
-    // `self_path'.
+    // If we got privileges with the fs set[ug]id bit, check that the privilege we
+    // got matches the one one we expected, ie that our effective uid/gid
+    // matches the uid/gid of `self_path`. This ensures that we were executed as
+    // `self_path', and not, say, as some other setuid program.
+    // We don't check that if we did not benefit from the set[ug]id bit, as
+    // can be the case in nosuid mounts or user namespaces.
     struct stat st;
     ASSERT(lstat(self_path, &st) != -1);
 
-    ASSERT(!(st.st_mode & S_ISUID) || (st.st_uid == geteuid()));
-    ASSERT(!(st.st_mode & S_ISGID) || (st.st_gid == getegid()));
+    ASSERT(!((st.st_mode & S_ISUID) && !no_suid) || (st.st_uid == geteuid()));
```

Comment:
Add a comment saying, this expresses:

```suggestion
   // if the binary is SGID and this is a privilege escalation, then the file's GID has to match the effective one (`egid`)
   ASSERT(!((st.st_mode & S_ISGID) && !didnt_gid) || (st.st_gid == egid));
```